File: src/jupyter_matlab_labextension/src/plugins/actions/baseAction.ts
// Copyright 2026 The MathWorks, Inc.

import { ICommunicationChannel } from '../matlabCommunication';

// Abstract Action class which other actions are meant to implement.

export abstract class BaseAction {
    /**
     * Whether {@link execute} should wait for the kernel's response before
     * resolving. When `true`, `execute` awaits an internal promise that is
     * settled by {@link onMsg} once the matching reply arrives; when `false`,
     * `execute` fires the request and returns without waiting.
     */
    protected blocking: boolean = false;

  /**
   * Sends this action's request to the kernel over the comm channel.
   *
   * Concrete actions validate `data`, send the request via `comm.send(...)`,
   * and — when {@link blocking} is `true` — wait for the kernel's reply
   * (delivered later through {@link onMsg}) before resolving.
   *
   * @param data - Action-specific payload to send to the kernel.
   * @param comm - Communication channel used to reach the kernel.
   */
  abstract execute(data: any, comm: ICommunicationChannel): void;

  /**
   * Handles a reply message received from the kernel for this action.
   *
   * Invoked by the comm layer's `onMsg` dispatcher when a message tagged with
   * this action type arrives. Implementations process the result (and, for
   * blocking actions, resolve or reject the promise awaited by
   * {@link execute}).
   *
   * @param data - Response payload sent by the kernel.
   * @param comm - Communication channel the message arrived on.
   */
  abstract onMsg(data: any, comm: ICommunicationChannel): void;

  /**
   * Returns the action type identifier for this action.
   *
   * Used to route requests and replies between the lab extension and the
   * kernel; the returned value is one of the {@link ActionTypes} constants.
   *
   * @returns The action type string.
   */
  abstract getActionName(): string;
}

File: src/jupyter_matlab_kernel/comms/labextension/actions/base_action.py
# Copyright 2026 The MathWorks, Inc.

from abc import ABC, abstractmethod


# Abstract Action class which other actions are meant to implement.
class ActionCommand(ABC):
    @abstractmethod
    async def execute(self, comm, data):
        """Runs the action and sends its result back to the lab extension.

        This is the single entry point invoked by the comm layer for a given
        action type. Concrete actions should validate ``data`` (typically via
        :meth:`validate_data`), perform their work, and report the outcome by
        calling ``comm.send(...)`` — both on success and on failure — rather
        than raising, so the lab extension always receives a response.

        Args:
            comm (ipykernel.comm.Comm): Communication channel used to send the
                result payload back to the lab extension.
            data (dict): Action-specific payload received from the lab
                extension.
        """
        pass

    def get_code(self, *args):
        """Builds the MATLAB code snippet this action needs to run in MATLAB.

        Only actions that drive MATLAB (e.g. opening a file in the editor)
        override this; actions that run purely on the kernel side leave it
        unimplemented.

        Args:
            *args: Action-specific values used to construct the code.

        Returns:
            str: The MATLAB code to execute.
        """
        pass

    def validate_data(self, data):
        """Validates the payload received from the lab extension.

        Concrete actions override this to assert that all required keys are
        present in ``data``, raising :class:`ValueError` when the payload is
        malformed so :meth:`execute` can report the error to the lab extension.

        Args:
            data (dict): Action-specific payload to validate.

        Raises:
            ValueError: If ``data`` is missing required fields.
        """
        pass
